### ISSUE TITLE
Bugs in CLiMF: compute sgds for items rated by user u

### ISSUE THREAD

**Issue by wenmin-wu** [guoguibing/librec#16]
According to the paper, the gradients of the objective for user i with respect to Vj is:
![image](https://cloud.githubusercontent.com/assets/9409333/8633548/27144f5c-2800-11e5-94f6-cf5ad6bd54fe.png)
This equation is different from that in the paper, due to the author has missed a Ui. From this equation, it's clear that the gradient of those items which haven't rated by user i is 0. However, the code given in this project is:

``` java
// compute sgds for items rated by user u
                Map<Integer, List<Double>> itemSgds = new HashMap<>();
                // for (int j : items) {
                for (int j = 0; j < numItems; j++) {

                    double fuj = predict(u, j);
                    List<Double> jSgds = new ArrayList<>();
                    for (int f = 0; f < numFactors; f++) {
                        double puf = P.get(u, f);
                        double qjf = Q.get(j, f);

                        double yuj = uv.contains(j) ? 1.0 : 0.0;
                        double sgd = yuj * g(-fuj) * puf - regI * qjf;

                        for (int k : uv.getIndex()) {
                            if (k == j)
                                continue;

                            double fuk = predict(u, k);
                            double x = fuk - fuj;

                            sgd += gd(-x) * (1.0 / (1 - g(x)) - 1.0 / (1 - g(-x))) * puf;
                        }
                        jSgds.add(sgd);
                    }

                    itemSgds.put(j, jSgds);
                }
```

I think the first `for` loop should be writen as: `for (int j : uv.getIndex())`.

**Comment by guoguibing** [comment]
@466152112 Can you take care of this issue first and possible provide a feedback? Thanks. I've been a bit busy recently.

**Comment by 466152112** [comment]
Ok, I will deal with this issue tomorrow.

**Comment by 466152112** [comment]
I read the original paper and find @wenmin-wu 's tips is right. In this way, the alg will be faster.

**Comment by 466152112** [comment]
Another issue I find is that:
when update for item j:
double x = fuk - fuj;
sgd += gd(x) * (1.0 / (1 - g(x)) - 1.0 / (1 - g(-x))) * puf;

From the paper, I think is 
double x = fuk - fuj;
sgd += gd(-x) * (1.0 / (1 - g(x)) - 1.0 / (1 - g(-x))) * puf;

The difference is the gd(-x) VS gd(x).  Whether I'm right? @wenmin-wu @guoguibing

**Comment by guoguibing** [comment]
Thanks for all your comments @wenmin-wu @466152112. I'll take your suggestion to fix it up.

**Comment by wenmin-wu** [comment]
@466152112 Sorry to reply late, a little busy recently. The gd is an even function, thus gd(-x) = gd(x). The original paper has no problem on this.

**Comment by wenmin-wu** [comment]
Another problem I have found in this implementation of CLiMF is the re-computation of ||Q||, The related code in this implementation equal to:
```java
double loss = 0;
for (int u = 0; u < numUsers; u++) {
	for (int i = 0; i < numItems; i++) {
		for (int f = 0; f < numFactors; f++) {
			double puf = P.get(u, f);
			double qjf = Q.get(j, f);
			loss += - 0.5 * (regU * puf * puf + regI * qjf * qjf);
		}
	}
}
```
As this result you have add -0.5 * numUsers *||Q|| to the loss, not ||Q||.

**Comment by yaronhe** [comment]
I think that the equation mentioned by
@wenmin-wu 
is incorrect since the author has put U_i out of the bracket.

In the released code of the author. He seems to remove both the bracket and U_i.

I am wondering why?
